swift/Lottie.swift

```diff
@@ -0,0 +1,152 @@
+        tvg_animation_get_duration(animation, &duration)
+
+        self.animation = animation
+        self.numberOfFrames = Int(numberOfFrames)
+        self.duration = CMTime(seconds: Double(duration), preferredTimescale: 600)
+    }
+
+    /// Renders a Lottie frame into a buffer.
+    /// - Parameters:
+    ///   - index: The index of the frame to render.
+    ///   - buffer: The buffer to render the frame into.
+    ///   - stride: The stride of the buffer.
+    ///   - size: The desired size of the rendered frame.
+    ///   - crop: Optional rectangle to crop the rendered frame.
+    public func render(frameAt index: Int, into buffer: Buffer, stride: Int, size: CGSize, crop: CGRect? = nil) throws {
+        guard index < numberOfFrames, index >= 0 else {
+            throw LottieError.frameIndexOutOfBounds
+        }
+
+        if canvas == nil {
+            canvas = try createCanvas(with: buffer, stride: stride, size: size)
+        }
+
+        prepareCanvasForRendering(frameIndex: index, size: size, crop: crop)
+        try renderCanvas()
+    }
+
+    /// Creates a new canvas for rendering, initializing the TVG engine and setting the target buffer.
+    private func createCanvas(with buffer: Buffer, stride: Int, size: CGSize) throws -> Canvas? {
+        guard tvg_engine_init(TVG_ENGINE_SW, 4) == TVG_RESULT_SUCCESS else {
+            throw LottieError.failedToInitialiseTvgEngine
+        }
+
+        let canvas = tvg_swcanvas_create()
+        tvg_swcanvas_set_target(canvas, buffer, UInt32(stride), UInt32(size.width), UInt32(size.height), TVG_COLORSPACE_ARGB8888)
+
+        let picture = tvg_animation_get_picture(animation)
+        tvg_canvas_push(canvas, picture)
+
+        return canvas
+    }
+
+    /// Prepares the canvas for rendering by setting the frame, adjusting the picture size, and optionally applying a crop.
+    private func prepareCanvasForRendering(frameIndex index: Int, size: CGSize, crop: CGRect?) {
+        tvg_canvas_clear(canvas, false, true)
+        tvg_animation_set_frame(animation, Float(index))
+
+        let picture = tvg_animation_get_picture(animation)
+        tvg_picture_set_size(picture, Float(size.width), Float(size.height))
+
+        if let crop {
+            let cropShape = tvg_shape_new()
+            tvg_shape_append_rect(
+                cropShape,
+                Float(crop.origin.x),
+                Float(crop.origin.y),
+                Float(crop.width),
+                Float(crop.height), 0, 0
+            )
+            tvg_paint_set_composite_method(picture, cropShape, TVG_COMPOSITE_METHOD_CLIP_PATH)
+        }
+
+        tvg_canvas_update_paint(canvas, picture)
+    }
+
+    /// Renders the prepared content of the canvas onto the actual canvas.
+    private func renderCanvas() throws {
+        guard tvg_canvas_draw(canvas) == TVG_RESULT_SUCCESS else {
+            throw LottieError.failedToDrawLottieFrameOntoCanvas
+        }
+
+        tvg_canvas_sync(canvas)
```

Comment:
> After getting the frame image, you can flush out all the pushed paint nodes from the canvas with Canvas::clear() call.

So shall we call `tvg_canvas_clear` here?

Reply:
Good find! I'll update the code to reflect that.